File: scripts/00-simulate_data.R
#### Preamble ####
# Purpose: Simulates a year of paramedic dispatch data
# Author: Muhammad Abdullah Motasim
# Date: 18 September 2024
# Contact: abdullah.motasim@mail.utoronto.ca
# License: MIT
# Pre-requisites: tidyverse, janitor, lubridate, truncnorm
# Any other information needed: None


#### Workspace setup ####
set.seed(1009067563)
library(tidyverse)
library(janitor)
library(lubridate)
library(truncnorm)

#### Simulate data ####

# There are around 27 000  entries for each year
# We will generate 30 000 entries to simulate data for the 2023 year
size <- 30000

# Create a sequence of dates and times within the 2023 year to choose from
datetime_sequence <- seq(from = as.POSIXct("2023-01-01 00:00"), 
                         to = as.POSIXct("2023-12-31 23:59"), 
                         by = "min")

# Format sequence to match data from Open Data Toronto
datetime_sequence <- format(datetime_sequence, format = "%Y-%m-%d %H:%M")

# Different calls have different priorities which are predetermined by the MDPS system
# A weighted sampling is used to account for this disparity 
priority_numbers = c(1, 3, 4, 5, 9, 11, 12, 13, 14)
priority_numbers_weights <- c(0.15, 0.2, 0.3, 0.2, 0.1, 0.05, 0.04, 0.03, 0.03)

# Different types of incidents occur more than others
# A weighted sampling is used to account for this disparity 
incident_types <- c("Medical", "Motor Vehicle Accident", "Emergency Transfer", 
                    "Fire", "Airport Standby", "Other")
incident_types_weights <- c(0.3, 0.3, 0.05, 0.1, 0.03, 0.2)

# There are a different amount of dispatchers sent for each call
# I assume there are typically 5 dispatchers sent per call
# We sample from a truncated normal distribution to account for this disparity 
max_dispactchers <- 22
min_dispactchers <- 0
mean <- 5
sd <- 3
  
  
simulated_data <-
  tibble(
    "id" = 1:size,
    
    "dispatch_time" = sample(x=datetime_sequence, size=size, replace=FALSE),

    "incident_type" = sample(x = incident_types, size = size, replace = TRUE, 
                             prob=incident_types_weights),
    
    "priority_number" = sample(x = priority_numbers, size=size, replace=TRUE, 
                               prob=priority_numbers_weights),
    
    "units_arrived_at_scene" = round(rtruncnorm(n = size, a = min_dispactchers,
                                                 b = max_dispactchers, mean = mean, sd = sd))
)

#### Write_csv
write_csv(simulated_data, file = "data/raw_data/simulated_ambulance_response_data.csv")

File: scripts/00-simulate_data.py
#### Preamble ####
# Purpose: Simulates a year of paramedic dispatch data (no scipy)
# Author: Muhammad Abdullah Motasim (translated to Python)
# Date: 18 September 2024

import pandas as pd
import numpy as np
from datetime import datetime

np.random.seed(1009067563)

#### Simulate data ####

size = 30000  # number of entries

# Create a datetime sequence for every minute in 2023
start = datetime(2023, 1, 1, 0, 0)
end = datetime(2023, 12, 31, 23, 59)
datetime_sequence = pd.date_range(start=start, end=end, freq="T").strftime("%Y-%m-%d %H:%M").tolist()

# Priority levels and weights
priority_numbers = [1, 3, 4, 5, 9, 11, 12, 13, 14]
priority_weights = [0.15, 0.2, 0.3, 0.2, 0.1, 0.05, 0.04, 0.03, 0.03]

# Incident types and weights
incident_types = ["Medical", "Motor Vehicle Accident", "Emergency Transfer",
                  "Fire", "Airport Standby", "Other"]
incident_weights = [0.3, 0.3, 0.05, 0.1, 0.03, 0.2]

# Simulate number of dispatchers using clipped normal distribution
mean = 5
sd = 3
min_dispatchers = 0
max_dispatchers = 22

dispatchers = np.random.normal(loc=mean, scale=sd, size=size)
dispatchers_clipped = np.clip(dispatchers, min_dispatchers, max_dispatchers).round().astype(int)

# Build the dataset
simulated_data = pd.DataFrame({
    "id": np.arange(1, size + 1),
    "dispatch_time": np.random.choice(datetime_sequence, size=size, replace=False),
    "incident_type": np.random.choice(incident_types, size=size, p=incident_weights),
    "priority_number": np.random.choice(priority_numbers, size=size, p=priority_weights),
    "units_arrived_at_scene": dispatchers_clipped
})

# Save to CSV
simulated_data.to_csv("data/raw_data/simulated_ambulance_response_data.csv", index=False)
